CIP-XXXX/README.md

```diff
@@ -37,236 +45,74 @@ The first approach allows for lower runtime resource consumption, but a general-
-
-##### Native Script by hash
-
-Request:
-
-- `ScriptHash`
-
-Response:
-
-- `NativeScript`
-
-##### Plutus Script by hash
-
-Request:
-
-- `ScriptHash`
-
-Response:
-
-- `PlutusScript`
-
-#### PlutusData
-
-##### By hash
-
-Request:
-
-- hash: `DataHash`
-
-Response:
-
-- `PlutusData`
-
-#### Reward history
-
-##### By stake address
-
-TODO
-
-#### Block
-
-##### By hash
-
-Request:
-
-- `BlockHash`
-
-Response:
-
-- `Block`
-
-#### Epoch
+### Transports
 
-TODO
+The API can be implemented across several transports. The goal is to allow several different clients, possibly written in different languages, to interact with wallets.
+For this reason we provide an [Openapi schema](./open-api.json), a [JSON-rcp](./json-rpc.json) schema, and an interface for an [injected Javascript](./ts-api.md) object in Typescript.
+We generate these interfaces from a high level specification of the endpoints [source](./link-to-ql-impl), ensuring that the information is consistent and easily updatable for different choices of transport layer.
 
-#### Genesis
+### Pagination
 
-TODO
+In CIP-30, pagination is not reliable, because there is no guarantee that the set of UTxOs does not change between calls. This behavior is not suitable for DeFi: consistency should be prioritized, and pagination should be avoided.
 
-#### Metadata
 
-##### By transaction hash
+### Handling of rollbacks
 
-TODO
+Transaction rollbacks are essential to blockchains: local node's view of the chain may be different from other nodes'. During conflict resolution, the node may issue a rollback event, that should be handled by dApps.
 
-#### Protocol parameters
+Customized chain followers, at least in principle, allow for "live" rollback handling: that is, a user-facing dApp can subscribe to a local view of a part of the UTxO set.
 
-##### By epoch
+General purpose query layers can also handle rollbacks just fine, but they don't propagate rollback events to dApps, because they do not possess any dApp-specific info to determine if a dApp *needs* to handle a particular rollback. dApps that work with general-purpose query layers follow pull-based architecture, rather than event subscription-based, which means they just request data as needed, instead of reacting to blockchain events.
 
-Request:
+In the context of this API, rollbacks should be acknowledged as a source of potential inconsistency between data pieces returned by different queries.
 
-- epoch: `UInt32`
+#### Error handling
 
-Response:
+Errors should be divided in two categories:
 
-- TODO: we should add `ProtocolParameters` to CIP-116
+- domain errors
+- transport errors (404, 500, etc)
 
-#### Stake pools
+Here we will only specify the domain errors. Users should also handle transport specific errors that can occur when interacting with the API.
 
-##### List of active stake pools
+##### Error Types
 
-Response:
+###### APIError
 
 ```
-{
-  pool_id: PoolPubKeyHash,
-  live_stake: UInt64,
-  active_stake: UInt64
-}[]
+APIErrorCode {
```

Comment:
@nrutledge both `Refused` and `AccountChange` can happen in this case. The errors  I list here would originate from the wallet, not the query layer provider. It could happen that for some reason the wallet disconnects while a user is waiting on a query, or that they switch accounts.